test/hotspot/jtreg/compiler/lib/compile_framework/CompileFramework.java

```diff
@@ -0,0 +1,254 @@
+                File asmtools = new File(dir, "asmtools.jar");
+                if (!asmtools.exists()) {
+                    throw new InternalCompileFrameworkException("Found jtreg.jar in classpath, but could not find asmtools.jar");
+                }
+                return asmtools.getAbsolutePath();
+            }
+        }
+        throw new InternalCompileFrameworkException("Could not find asmtools because could not find jtreg.jar in classpath");
+    }
+
+    private static void compileJavaSources(String sourceDir, List<SourceCode> javaSources) {
+        if (javaSources.size() == 0) {
+            System.out.println("No java sources to compile.");
+            return;
+        }
+        System.out.println("Compiling Java sources: " + javaSources.size());
+
+        List<String> javaFileNames = writeSourcesToFile(sourceDir, javaSources);
+        compileJavaFiles(javaFileNames);
+
+        System.out.println("Java sources compiled.");
+    }
+
+    private static void compileJavaFiles(List<String> javaFileNames) {
+        // Compile JAVA files with javac, in the "compile.jdk".
+        List<String> command = new ArrayList<>();
+
+        command.add("%s/bin/javac".formatted(System.getProperty("compile.jdk")));
+        command.add("-classpath");
+        command.add(System.getProperty("java.class.path"));
+        command.add("-d");
+        command.add(System.getProperty("test.classes"));
+        for (String fileName : javaFileNames) {
+            command.add(new File(fileName).getAbsolutePath());
+        }
+
+        executeCompileCommand(command);
+    }
+
+    private static List<String> writeSourcesToFile(String sourceDir, List<SourceCode> sources) {
+        List<String> storedFiles = new ArrayList<String>();
+        for (SourceCode sourceCode : sources) {
+            String extension = sourceCode.kind.name().toLowerCase();
+            String fileName = sourceDir + "/" + sourceCode.className.replace('.','/') + "." + extension;
+            writeCodeToFile(sourceCode.code, fileName);
+            storedFiles.add(fileName);
+        }
+        return storedFiles;
+    }
+
+    private static void writeCodeToFile(String code, String fileName) {
+        System.out.println("File: " + fileName);
+        File file = new File(fileName);
+        File dir = file.getAbsoluteFile().getParentFile();
+        if (!dir.exists()){
+            dir.mkdirs();
+        }
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
+            writer.write(code);
+        } catch (Exception e) {
+            throw new CompileFrameworkException("Could not write file: " + fileName, e);
+        }
+    }
+
+    private static void executeCompileCommand(List<String> command) {
+        System.out.println("Compile command: " + String.join(" ", command));
+
+        ProcessBuilder builder = new ProcessBuilder(command);
+        builder.redirectErrorStream(true);
```

Comment:
Same streams question. Not always we'd like to have errors in the `stderr`

Reply:
But here I think I want everything to go to the `stdout`. That way, I can check below that the compilation had neither any errors nor warnings. See `output.equals("")`.

Reply:
> See output.equals("").
Well, that doesn't require redirection. You check the process' output, not the host (JTReg) output. 

> But here I think I want everything to go to the stdout.
Well, here I'd like to call the same argument of volume. If we generate 100s of files, using not the perfect generators (that generate warnings, can create semi-obsolete code, etc.)... we'd not want to stare at their outputs manually.

Reply:
I'm not sure I understand what you are saying.

`redirectErrorStream`: Tells whether this process builder merges standard error and standard output.

So all I'm doing is merging the stdout and stderr from the process. Now everything from the process goes to the stdout of the process, right? Then I can check `output.equals("")` and that captures that there is neither any thing on the stdout nor on the stderr of the process.

Nothing is actually being printed in the JTREG stdout. Correct?

Reply:
BTW: you put not just my text as quotation but also your reply ;)

Reply:
> So all I'm doing is merging the stdout and stderr from the process.
..
> Nothing is actually being printed in the JTREG stdout. Correct?

You're right. Ignore that comment then.